text/2017-11-02-guards.md

```diff
@@ -0,0 +1,316 @@
+
+```rust
+#[cold]
+unsafe fn resize(&self, new_cap: usize) {
+    // Load the bottom, top, and buffer.
+    let b = self.bottom.load(Relaxed);
+    let t = self.top.load(Relaxed);
+
+    let buffer = self.buffer.load(Relaxed, epoch::unprotected());
+
+    // Allocate a new buffer.
+    let new = Buffer::new(new_cap);
+
+    // Copy data from the old buffer to the new one.
+    let mut i = t;
+    while i != b {
+        ptr::copy_nonoverlapping(buffer.deref().at(i), new.at(i), 1);
+        i = i.wrapping_add(1);
+    }
+
+    let guard = &epoch::pin();
+
+    // Replace the old buffer with the new one.
+    let old = self.buffer
+        .swap(Owned::new(new).into_ptr(guard), Release, guard);
+
+    // Destroy the old buffer later.
+    guard.defer(move || old.into_owned());
+
+    // If the buffer is very large, then flush the thread-local garbage in order to
+    // deallocate it as soon as possible.
+    if mem::size_of::<T>() * new_cap >= FLUSH_THRESHOLD_BYTES {
+        guard.flush();
+    }
+}
+```
+</details>
+
+First, `self.buffer` is loaded using an unprotected guard, and after
+that it is loaded again (using `swap`) with a real guard. If we were
+to create two guards on the stack, perhaps it would be easy to later
+mistake one for another.
+
+## Changes to the `Collector` interface
+
+### Should `Handle` be `Send`?
+
+If we allow `Handle`s to be `Send`, the following code will compile:
+
+```rust
+let c = Collector::new();
+let h = c.handle();
+let guard = h.pin();
+
+thread::spawn(move || {
+    let guard = h.pin();
+});
+```
+
+This is a suspicious, although not necessarily a *wrong* piece of code. However, if we
+allowed this to be possible, `guard_count` would have to be an atomic integer, and
+atomic operations on the counter would unnecessarily slow down pinning.
+
+For that reason, `Handle` will not be `Send`.
+
+Note that, currently, `Handle::clone` creates a brand new handle that can be passed to
+another thread (see [PR #26](https://github.com/crossbeam-rs/crossbeam-epoch/pull/26)),
+but with non-`Send` handles that cloning behavior is not very useful anymore.
+
+### Accessing the default handle
+
+Moreover, we'been having difficulties finding a satisfying signature for the
+`default_handle` method (see [PR #28](https://github.com/crossbeam-rs/crossbeam-epoch/pull/28)).
+The dilemma is whether we should have:
+
+1. Unsafe function `default_handle` returning a `&'static Handle`.
+2. Safe function `with_default_handle` taking a closure (similar to `LocalKey::with`).
+   This is slightly unergonomic.
+3. Safe function `default_handle` that uses reference counting (`Arc`) to return a
+   reference to the default handle. This would be very similar to the behavior of
+   `std::thread::current()`. The main problem is that cloning an `Arc` would be slow.
+
+But if handles are not `Send`, a fourth option emerges:
+
+4. Same as the third option, but have internal reference counting using `Rc` instead of `Arc`.
```

Comment:
I think the options 3. and 4. are as unsafe as 1. and 2., as they may panic: https://doc.rust-lang.org/1.21.0/std/thread/struct.LocalKey.html#panics . For this reason, I think the signature of `thread_local!`'s `HANDLE.with()` is misleading: while the signature says it's safe, it is actually quite unsafe.  (I'd like to cc @joshlf, who met this panic in the development of elfmalloc.)

That being said, I'd like to advocate for option 1., which doesn't incur runtime performance.

Reply:
I've rewritten this section of the text. Hopefully it's clearer now and objectively lays out the pros & cons of each option. I've also removed the 3. option.

> For this reason, I think the signature of thread_local!'s HANDLE.with() is misleading: while the signature says it's safe, it is actually quite unsafe. 

Just to be clear: when you say "it's actually quite unsafe", do you mean "it's actually quite dangerous", or are you talking about "unsafe" as in "invokes undefined behavior"?